src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,66 +191,182 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
+    // constructor parameter: Array[target, mh1, mh2, ...]
+    private static byte[] spin(ClassDesc ifaceDesc, String name, List<LocalMethodInfo> methods) {
+        ClassDesc proxyDesc = ifaceDesc.nested("$MethodHandleProxy");
+        return Classfile.build(proxyDesc, clb -> {
+            clb.withSuperclass(CD_Object);
+            clb.withFlags(ACC_FINAL | ACC_SYNTHETIC);
+            clb.withInterfaceSymbols(ifaceDesc);
+            clb.with(RuntimeVisibleAnnotationsAttribute.of(Annotation.of(CD_WrapperInstance,
+                    AnnotationElement.of("implementedType", AnnotationValue.ofClass(ifaceDesc)))));
+            // <init>
+            clb.withMethodBody(INIT_NAME, MTD_void_MethodHandleArray, 0, cob -> {
+                cob.aload(0);
+                cob.dup();
+                cob.invokespecial(CD_Object, INIT_NAME, MTD_void);
+                // target field
+                cob.dup();
+                cob.aload(1);
+                cob.iconst_0();
+                cob.aaload();
+                cob.putfield(proxyDesc, TARGET_FIELD_NAME, CD_MethodHandle);
+
+                // implemented methods
+                int i = 1;
+                for (var method : methods) {
+                    cob.dup();
+                    cob.aload(1);
+                    cob.constantInstruction(i++);
+                    cob.aaload();
+                    cob.putfield(proxyDesc, method.fieldName, CD_MethodHandle);
                 }
+
+                cob.pop();
+                cob.return_();
             });
-        } else {
-            proxy = Proxy.newProxyInstance(proxyLoader,
-                                           new Class<?>[]{ intfc, WrapperInstance.class },
-                                           ih);
-        }
-        return intfc.cast(proxy);
-    }
 
+            // wrapper instance methods
+            clb.withField(TARGET_FIELD_NAME, CD_MethodHandle, ACC_PRIVATE | ACC_FINAL);
+            clb.withMethodBody(WRAPPER_INSTANCE_TARGET_NAME, MTD_MethodHandle, ACC_PRIVATE, cob -> cob
+                    .aload(0)
+                    .getfield(proxyDesc, TARGET_FIELD_NAME, CD_MethodHandle)
+                    .areturn());
+
+            // object methods
+            clb.withMethodBody("toString", MTD_String, ACC_PUBLIC, cob -> cob
+                    .aload(0)
+                    .invokestatic(CD_Objects, "toIdentityString", MTD_String_Object)
+                    .areturn());
+            clb.withMethodBody("hashCode", MTD_int, ACC_PUBLIC, cob -> cob
+                    .aload(0)
+                    .invokestatic(CD_System, "identityHashCode", MTD_int_Object)
+                    .ireturn());
+            clb.withMethodBody("equals", MTD_boolean_Object, ACC_PUBLIC, cob -> cob
+                    .aload(0)
+                    .aload(1)
+                    .ifThenElse(Opcode.IF_ACMPEQ, CodeBuilder::iconst_1, CodeBuilder::iconst_0)
+                    .ireturn());
```

Comment:
The object methods could be left untouched, as the class inherits the default implementation from `java.lang.Object`. Interfaces can't have default methods with those signatures.

Reply:
Do interfaces still implement these properly without definition if they extend interfaces that declare `@Override boolean equals();` etc.?

Reply:
Yes. You can use [`Comperator<T>`][1]  as test.

[1]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Comparator.html#equals(java.lang.Object)

Reply:
Added a test and confirmed they work in the latest patch.